handler.go

```diff
@@ -0,0 +1,214 @@
+	)
+	defer span.End()
+	var reply *dhcpv4.DHCPv4
+	switch mt := m.MessageType(); mt {
+	case dhcpv4.MessageTypeDiscover:
+		reply = s.handleDiscover(ctx, tracer, m)
+	case dhcpv4.MessageTypeRequest:
+		reply = s.handleRequest(ctx, tracer, m)
+	case dhcpv4.MessageTypeRelease:
+		s.handleRelease(ctx, m)
+	default:
+		log.Info("received unknown message type", "type", mt)
+	}
+	c := codes.Ok
+	cmsg := fmt.Sprintf("no reply send DHCP: %v", m.MessageType())
+	if reply != nil {
+		span.SetAttributes(attribute.String("IP", reply.YourIPAddr.String()))
+		if _, err := conn.WriteTo(reply.ToBytes(), peer); err != nil {
+			log.Error(err, "failed to send DHCP")
+			c = codes.Error
+			cmsg = fmt.Sprintf("failed to send DHCP: %v, err: %v", reply.MessageType().String(), err.Error())
+		} else {
+			log.Info("sent DHCP response")
+			cmsg = fmt.Sprintf("sent DHCP: %s", reply.MessageType().String())
+		}
+	}
+	span.SetStatus(c, cmsg)
+}
+
+// handleDiscover handles DHCP packets message type of discover.
+func (s *Server) handleDiscover(ctx context.Context, tracer trace.Tracer, m *dhcpv4.DHCPv4) *dhcpv4.DHCPv4 {
+	log := s.Log.WithValues("mac", m.ClientHWAddr.String())
+	ctx, span := tracer.Start(ctx, "DHCP Discover Message")
+	log.Info("received discover packet")
+	d, n, err := s.Backend.Read(ctx, m.ClientHWAddr)
+	if err != nil {
+		s.Log.Info("not sending DHCP OFFER", "mac", m.ClientHWAddr, "error", err)
+		span.SetStatus(codes.Error, fmt.Sprintf("not sending DHCP OFFER: %v", err))
+		span.End()
+		return nil
+	}
+	span.SetAttributes(
+		attribute.String("Subnet", d.SubnetMask.String()),
+		attribute.String("DefaultGateway", d.DefaultGateway.String()),
+		attribute.String("Nameservers", fmt.Sprintf("%v", d.NameServers)),
+		attribute.String("Hostname", d.Hostname),
+		attribute.String("DomainName", d.DomainName),
+		attribute.Int("LeaseTime", int(d.LeaseTime)),
+		attribute.Bool("AllowNetboot", n.AllowNetboot),
+		attribute.String("IpxeScriptURL", fmt.Sprintf("%v", n.IpxeScriptURL)),
+	)
+	span.SetStatus(codes.Ok, "done reading from backend")
+	span.End()
+
+	mods := []dhcpv4.Modifier{
+		dhcpv4.WithMessageType(dhcpv4.MessageTypeOffer),
+		dhcpv4.WithGeneric(dhcpv4.OptionServerIdentifier, s.IPAddr.IPAddr().IP),
+		dhcpv4.WithServerIP(s.IPAddr.IPAddr().IP),
+	}
+	mods = append(mods, s.setDHCPOpts(ctx, m, d)...)
+
+	if s.NetbootEnabled && s.isNetbootClient(m) {
+		mods = append(mods, s.setNetworkBootOpts(ctx, m, n))
+	}
+	reply, err := dhcpv4.NewReplyFromRequest(m, mods...)
+	if err != nil {
+		return nil
+	}
+	log.Info("sending offer packet")
+	return reply
+}
+
+// handleRequest handles DHCP packets message type of request.
+func (s *Server) handleRequest(ctx context.Context, tracer trace.Tracer, m *dhcpv4.DHCPv4) *dhcpv4.DHCPv4 {
+	log := s.Log.WithValues("mac", m.ClientHWAddr.String())
+	ctx, span := tracer.Start(ctx, "DHCP Request Message")
+	log.Info("received request packet")
+	d, n, err := s.Backend.Read(ctx, m.ClientHWAddr)
+	if err != nil {
```

Comment:
In this code path I don't think span.End() gets called at all.
I think the pattern of using defer would be preferable.

Reply:
my understanding of otel, admittedly very shallow, is that we start and end spans around functionality we want to be timed. I don't know if that is correct, but that is the intention here. `s.Backend.Read` will typically contain some kind of network call. Moving to a `defer span.End()` would change the timing to include more than just the underlining network call of `s.Backend.Read`. Is that ok? or "the way" to do otel? I'm definitely open to changing this. what do you think?